draft/ZEP0010.md

```diff
@@ -0,0 +1,413 @@
+``extensions`` container. This ZEP is putting in place the mechanism so the
+community can experiment with such extensions before their standardization.
+
+#### Offset (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "array",
+    ...,
+    "extensions": [
+        {
+            "name": "example.offset",
+            "configuration": { "offset": [ 12, 24 ] }
+        }
+    ]
+}
+```
+
+The ``example.offset`` extension contains an array of the same order as the
+shape of the containing array specifying which element of the array should be
+considered as the origin, e.g., `[0, 0]`. This allows the reuse of subregions
+of an array without the need to rewrite the data.
+
+Note that in this example of the extension is ``must_understand=true`` meaning
+an implementation which does not support the ``example.offset`` extension
+should raise an error.
+
+#### Statistics (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "array",
+    ...,
+    "extensions": [
+        {
+            "name": "example.array-statistics",
+            "must_understand": false,
+            "configuration": {
+                "min": 5,
+                "max": 1023
+            }
+        }
+    ]
+}
+```
+
+The ``example.array-statistics`` extension contains two fields -- ``min``
+and ``max`` specifying the range of values which are present in the array,
+reducing the need to read every byte. ``must_understand`` is false, so
+implementations can safely ignored the extension.
+
+#### Skip empty chunks (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    ...,
+    "extensions": [
+        "example.skip_empty_chunks"
+    ]
+}
+```
+
+Currently the "write_empty_chunks" flag in zarr-python is not propagated
+to the zarr.json file. An extension like ``example.skip_empty_chunks``
+could serve as a no-configuration flag in the metadata to inform
+implementations that empty chunks should not be written.
+
+
+
+#### Domain metadata (group)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "group",
+    ...,
+    "extensions": [
+        {
+            "name": "example.domain-metadata",
+            "must_understand": false,
+            "configuration": {
+                "multiscale": {
+                    "datasets": [
+                        "path/to/array/1",
+                        "path/to/array/2",
+                        "path/to/array/3"
+                    ]
+                }
+            }
+        }
+    ],
+}
+```
+
+Domain-specific metadata is introduced in the ``example.domain-metadata``
+extension which allows encoding a relationship between multiple arrays at the
+group level. Here, a "multiscale pyramid" of arrays is being defined which is
+a common idiom in the geospatial and bioimaging uses of Zarr.
+
+The value of specifying this metadata as an extension as opposed to a
+user-attribute is the clear registration of the extension name, providing a
```

Comment:
what is the link between registering an extension name and putting the metadata associated with that name in a new field in array metadata? Why could the registered names be used at the top-level of `attributes`?